more type info

dragazo opened the pull request:
In the python client I'm building, I'm using type info from the services metadata to make type annotations. This works great for the standard types built into `input-types.js`, but it's not future-proofed for adding new types via `defineType`, for which I have to fall back to `Any`.

This PR adds an extra argument to `defineType` to denote the base type, and we expose this metadata through a new endpoint at `/services/input-types/base-types`. It's meant to give the simplified representation type for a specialized type. `Any` is available as a base type for anything that doesn't have an obvious base.

The base type must either be a string or an object of the same format given by the input type parser. This is a bit verbose for specifying new enum types, so I've also added `defineEnum` to shorthand that process.

Note: `baseType` is required, and `defineType` ensures that it exists.

dragazo commented:
The most recent commit fixes merge conflicts and refactors the new base type system I added initially. Now, all types are introduced with `defineType`, even the ones defined in the the `input-types.js` file itself (with the sole exception of `Any`, which is basetype-less). The endpoint is changed to just `services/input-types`, and for future proofing it now returns an object like `{baseTypes}` instead of `baseTypes` directly (so we can add extra info under different fields later on)

brollb reviewed (commented):
A couple comments that don't necessarily need to be addressed here but are important to keep in mind when refactoring the type system (especially since we are changing the public REST API):
- It would be nice to generate documentation for types to host at `/docs`. This is especially true when we are working with large enums as all possible values can get cumbersome and difficult to read from the error text in NetsBlox. Base types could be helpful here if we have programmatic ways of generating the docs.
- We should probably display the types (at least in the documentation) using the Snap names for them (eg, `Array` should be `List`).
- Other types, like `SerializedFunction`, probably need a display name, too. It is important that it is serialized but that doesn't matter to the end user.
- Eventually, it would be nice to scope custom types to the service that uses them. This would allow us to generate type docs for the specific service (like the many used by PhoneIoT).

A couple specific questions:
- There is something I don't like about how the base types are handled; they aren't really inheriting anything and are more like categories or metadata (since they can have a label and parameters). Currently, I could make a new type and set the baseType to an arbitrary (existing) type and it wouldn't really mean anything. Nor would it actually benefit me in any obvious way. The only exception is that if it is an enum, it would help with the docs if I pass the parameters to the base type but they aren't used anywhere... Would it be better to actually have the types call the base type first?
- This also might make doc generation for types easier as each type could possible have a template for its docs which have access to the `super` type's doc generator, base type's parameters, and fallback on the base type's doc generator.
- It might be nice to change the REST API to just return a list of input types that are supported. These could later be extended to support the comments from earlier. As an example:
```javascript
{
  name: "ButtonStyle",
  base: "Enum",
  params: ["rectangle", "ellipse", ... "circle"],
  service: "PhoneIoT",
  docs: "https://editor.netsblox.org/docs/services/PhoneIoT/types.html#ButtonStyle"
}
```

dragazo commented:
Ok, `defineType` now has a big refactor that allows more info to be specified. It now takes an sanitized info object containing things like `displayName` and `description`. It also composes the parser function with the base type automatically, which lets us do things like:

```js
defineType({
    name: 'BoundedInteger',
    description: 'An Integer with a minimum and/or maximum value.',
    baseType: 'BoundedNumber',
    baseParams: p => p, // pass our params to the base type parser
    parser: input => { // we're guaranteed this is a (bounded) number because of base type
        if (!Number.isInteger(input)) throw new InputTypeError('Number must be an integer (whole number)');
        return input;
    },
});
```

Minor inconvenience: for the composition to work in general, I needed to make the derived parser async in case the base parser is async, so all type parsing (other than `Any`, which is no-op) has to be wrapped in await. I already fixed this everywhere I saw it was a problem, and updated the unit tests.

Currently, I was planning on keeping the types metadata separate from services metadata to avoid duplicated json if a type is used multiple types (I kinda just wanted a centralized types area and then other things point into it by name). But if nothing else, there might be a better endpoint for it than the one I used. Actually, I would've prefered to have this be a separate field returned by the main `/services` endpoint, but since that's an array we can't add extra info now (which is why I made the input types endpoint return an object that currently only has one field).

brollb commented on src/server/services/procedures/alexa/types.js:
```diff
@@ -159,7 +159,11 @@ const SlotTypes = [
     'AMAZON.WrittenCreativeWorkType',
 ];
 
+let _registeredTypes = false;
```
Is this added just in case or is it actually registering the types twice?

dragazo replied:
I'm not sure why, but `npm test` was suddenly failing there for some reason, and that fixed it

brollb commented:
Overall, it looks good. Another comment: what other fields might we want to add to the `input-types` endpoint? I actually prefer it returning a list as it seems any extra metadata could be stored within each type (or on another endpoint). However, I might be overlooking some motivating example.

dragazo commented:
For the extra fields thing, I can't think of anything so far - was just a future proofing thing. But you're probably right, any other info could almost certainly just be in the data for each type. I'll swap it over to an array.
